### PR TITLE
vm: restore JSON tests for NEO3

### LINKED ISSUE #892: Restore testing with neo-vm test set for version 3
#858 temporarily disabled neo-vm tests, after fixing #873, #875, #876, #877, #878, #879, #880, #881, #882, #883 and #884 we should restore this testing, but it would also require some update to the system we use as tests have changed.

### PR DESCRIPTION
**Pull request by fyrchik**
Enough work has been done in VM to support NEO3 set of tests.
Minor fixes were done.
Closes #892 .

### CODE REVIEW

File: pkg/vm/json_test.go [hunk 1]
```diff
@@ -90,6 +92,30 @@ const (
...
+	stringToOpcode["CALL_L"] = opcode.CALLL
+	stringToOpcode["JMPEQ_L"] = opcode.JMPEQL
+	stringToOpcode["JMPNE_L"] = opcode.JMPNEL
+	stringToOpcode["JMPGE_L"] = opcode.JMPGEL
+	stringToOpcode["JMPGT_L"] = opcode.JMPGTL
+	stringToOpcode["JMPLE_L"] = opcode.JMPLEL
+	stringToOpcode["JMPLT_L"] = opcode.JMPLTL
+	stringToOpcode["JMPIFNOT_L"] = opcode.JMPIFNOTL
+	stringToOpcode["JMPIF_L"] = opcode.JMPIFL
+	stringToOpcode["JMP_L"] = opcode.JMPL
```
Comment (roman-khimov) [inline pkg/vm/json_test.go:119]: It's better to add proper `FromString(string) Opcode` to `opcode` package. It might be useful in other places. And we can also change `String()` to use underscored versions.

File: pkg/vm/json_test.go [hunk 2]
```diff
@@ -291,12 +317,48 @@ func (v *vmUTState) UnmarshalJSON(data []byte) error {
...
 func (v *vmUTScript) UnmarshalJSON(data []byte) error {
-	b, err := decodeBytes(data)
-	if err != nil {
+	var ops []string
+	if err := json.Unmarshal(data, &ops); err != nil {
 		return err
 	}
 
-	*v = vmUTScript(b)
+	decodeSingle := func(s string) ([]byte, bool) {
```
Comment (roman-khimov) [inline pkg/vm/json_test.go:356]: It can a proper function, I guess.

File: pkg/vm/json_test.go [hunk 3]
```diff
@@ -303,12 +303,12 @@ func execStep(t *testing.T, v *VM, step vmUTStep) {
...
-	case `"Break"`:
+	switch s := string(data); strings.ToUpper(s) {
+	case `"BREAK"`:
 		*v = vmUTState(breakState)
-	case `"Fault"`:
+	case `"FAULT"`:
 		*v = vmUTState(faultState)
-	case `"Halt"`:
+	case `"HALT"`:
 		*v = vmUTState(haltState)
```
Comment (roman-khimov) [inline pkg/vm/json_test.go:331]: We have proper `(*State).UnmarshalJSON`, why not use it here?

File: pkg/vm/json_test.go [hunk 4]
```diff
@@ -150,8 +149,17 @@ func testFile(t *testing.T, filename string) {
 	data, err := ioutil.ReadFile(filename)
 	require.NoError(t, err)
 
+	// get rid of possible BOM
+	if len(data) > 2 && data[0] == 0xef && data[1] == 0xbb && data[2] == 0xbf {
+		data = data[3:]
+	}
+	if strings.HasSuffix(filename, "MEMCPY.json") {
+		return // FIXME this file is not a valid JSON
```
Comment (roman-khimov) [inline pkg/vm/json_test.go:157]: There should be a neo-vm bug for it.

File: pkg/vm/json_test.go (hunk 1 above)
Comment (fyrchik) [reply pkg/vm/json_test.go:119]: `String` is autogenerated, so we need to rename all constants to use underscore. This was done and declined in #907 or #901, because it makes linter unhappy.
And `FromString` is not that useful without corresponding `String` change. I used it only here, so decided not to add it to the opcode, but this is easy.

File: pkg/vm/json_test.go (hunk 3 above)
Comment (fyrchik) [reply pkg/vm/json_test.go]: Fixed.

File: pkg/vm/json_test.go (hunk 4 above)
Comment (fyrchik) [reply pkg/vm/json_test.go:155]: https://github.com/neo-project/neo-vm/issues/322

File: pkg/vm/json_test.go (hunk 2 above)
Comment (fyrchik) [reply pkg/vm/json_test.go:343]: Fixed.

File: pkg/vm/json_test.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/vm/json_test.go:117]: Changing constants is not an option, but changing their string representations is. Not autogenerated, but given that nothing autogenerates `FromString`...

Comment (fyrchik) [reply pkg/vm/json_test.go]: Fixed.

File: pkg/vm/opcode/from_string.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package opcode
+
+import "errors"
+
+var stringToOpcode = make(map[string]Opcode)
+
+func init() {
+	for i := 0; i < 255; i++ {
+		op := Opcode(i)
+		stringToOpcode[op.String()] = op
```
Comment (roman-khimov) [inline pkg/vm/opcode/from_string.go:10]: There are some invalid opcodes and you're generating an `Opcode(XXX)` map for them, I don't think that's correct.

File: pkg/vm/json_test.go [hunk 6]
```diff
@@ -291,6 +293,14 @@ func execStep(t *testing.T, v *VM, step vmUTStep) {
 	}
 }
 
+func fromString(s string) StackItem {
```
Comment (roman-khimov) [inline pkg/vm/json_test.go:310]: `vm.fromString` is waaay too ambigous name.

File: pkg/vm/vm.go [hunk 7]
```diff
@@ -376,14 +376,14 @@ func (v *VM) StepInto() error {
...
-		v.state |= haltState
+		v.state = haltState
 	}
 
 	if v.HasStopped() {
 		return nil
 	}
 
-	if ctx != nil && ctx.prog != nil {
+	if ctx != nil && len(ctx.prog) != 0 {
```
Comment (roman-khimov) [inline pkg/vm/vm.go:386]: Doesn't seem to be related to `State` changes, why do you need that?

File: pkg/vm/vm_test.go [hunk 8]
```diff
@@ -2435,7 +2435,9 @@ func makeProgram(opcodes ...opcode.Opcode) []byte {
 
 func load(prog []byte) *VM {
 	vm := New()
-	vm.LoadScript(prog)
+	if len(prog) != 0 {
```
Comment (roman-khimov) [inline pkg/vm/vm_test.go:2438]: What's the use case?

Comment (fyrchik) [reply pkg/vm/vm_test.go:2438]: https://github.com/neo-project/neo-vm/blob/master/tests/neo-vm.Tests/Tests/Others/OtherCases.json#L22

Comment (roman-khimov) [reply pkg/vm/vm_test.go:2438]: Well, you'd better write it in commit message then. Looks weird anyway.

File: pkg/vm/opcode/from_string.go (hunk 5 above)
Comment (fyrchik) [reply pkg/vm/opcode/from_string.go:10]: I thought about this and see no big problem here: the string must be precisely defined (we don't return opcode for any random string), it is nice to have `FromString(op.String()) == op` as well as `FromString(s).String() == s`, and if really needed, it is better to have a separate function to check if opcode is implemented by our VM.
